pagopa/io-app
# chore(IDPay): [IODPAY-177] Add ability to enroll multiple instrument concurrently

### ⚠️ This PR depends on #4416

## Short description
This PR introduces a new mechanism for simultaneously handling multiple requests for enrolling or deactivating instruments, without blocking the user interface. Each response to a request is then handled individually. Currently, the machine that manages the enrollment process does not permit the sending of multiple requests.


https://user-images.githubusercontent.com/6160324/230426448-9a0ef8a7-63ad-48e2-8b65-2f3bfbc963cd.mp4



## List of changes proposed in this pull request

- [ts/features/idpay/initiative/configuration/xstate/actions.ts](https://github.com/pagopa/io-app/pull/4471/files#diff-1af1026b4e387e27a8b35f3923bfeb5400e10f3b01e86a831ad5df8ebee9d0d7)
  - Added `showInstrumentFailureToast` to display a toast in case of enrollment/deactivation failures
- [ts/features/idpay/initiative/configuration/xstate/context.ts](https://github.com/pagopa/io-app/pull/4471/files#diff-b329754fa14af94879e47deb3b0ccc99b9aeb76ae8ef0c7d5edc72af6b1796a5)
  - Removed unused values
- [ts/features/idpay/initiative/configuration/xstate/events.ts](https://github.com/pagopa/io-app/pull/4471/files#diff-22cbb26b3f180921c8f7d1d86041c6270e1999f2f26163896a19ee3d80ad5c9b)
  - Added events to handle enrollment/deactivation: `ENROLL_INSTRUMENT_SUCCESS`, `ENROLL_INSTRUMENT_FAILURE`, `DELETE_INSTRUMENT_SUCCESS`, `DELETE_INSTRUMENT_FAILURE`
  - Removed `STAGE_INSTRUMENT` event, instrument staging is now handled by the components
- [ts/features/idpay/initiative/configuration/xstate/selectors.ts](https://github.com/pagopa/io-app/pull/4471/files#diff-1710a733b11722d3dc9d71f3a66d8a7529c9446ca85cdedb94be50ee83d17d1f):
  - Modified selector to match new machine context structure
- [ts/features/idpay/initiative/configuration/xstate/services.ts](https://github.com/pagopa/io-app/pull/4471/files#diff-2ef56dfac7a99de77a9ee8f488fe7e6b4093973d642207faf9d593da561a2528)
  - Added a new service, `instrumentsEnrollmentService`, which can handle async events from parent machine
- [ts/features/idpay/initiative/configuration/xstate/machine.ts](https://github.com/pagopa/io-app/pull/4471/files#diff-d166bc2cc43785575c9570de3a3c577e685a12b41e5028f6050d03613cfd5419)
  - Removed unused actions
  - Removed `ENROLLING_INSTRUMENT` and `DELETING_INSTRUMENT` statuses
  - Added new actions to manage `instrumentsEnrollmentService`
- [ts/features/idpay/initiative/configuration/components/InstrumentEnrollmentSwitch.tsx](https://github.com/pagopa/io-app/pull/4471/files#diff-42b37f8c786239c06e0144ac6b02bce2a70ceb54f32197c486101bd2b8848722)
- [ts/features/idpay/initiative/configuration/screens/InstrumentsEnrollmentScreen.tsx](https://github.com/pagopa/io-app/pull/4471/files#diff-e57ad3d99b7dbdc319850166ea2e2c79d356a9c9ebfa3606c0ed51d139289157)
  - Removed `STAGE_INSTRUMENT` event login, instrument staging is now handled by the components
  
## How to test
- With the `io-dev-server`, set a 2000-3000 ms delay with the `config.json`. This will help testing;
- Go to the wallet section of the app and make sure to have at least two cards added;
- Start a new initiative configuration until the instrument configuration step;
- Here you should be able to enroll multiple instruments simultaneously.
